### PR TITLE
feat: add state field and archivePromotion endpoint

### LINKED ISSUE #6655: Promotion records should have a "state" field that relates to the current status of the promotion
The currently defined states would:

created - The default status on creation
active - When the current promotion is active
completed - When the promotion has completed it's run but has not been archived
archived - When the user wishes to be able to "hide" the promotion, they can place it in the archived state

### PR DESCRIPTION
**Pull request by brent-hoover**
Resolves #6655
Impact: **minor**
Type: **feature**

Adds the "state" field and the `archivePromotion` endpoint. Also includes a "watcher" that marks records either "active" or "completed" so that shop TZ specific math can be done here. Also a utility to get times based on shop timezone with in-memory caching to avoid constantly hitting the Shops collection.

### CODE REVIEW

File: packages/api-plugin-promotions/src/resolvers/Mutation/archivePromotion.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * @method archivePromotion
+ * @summary Mark a promotion as archived
+ * @param {Object} _ - unused
+ * @param {Object} args - The input arguments
+ * @param {Object} args.input - the promotionId of the promotion to archive
+ * @param {Object} context - an object containing the per-request state
+ * @return {Promise<Object>} archiveProduct payload
+ */
+export default async function updatePromotion(_, { input }, context) {
```
Comment (vanpho93) [inline packages/api-plugin-promotions/src/resolvers/Mutation/archivePromotion.js]: function name or filename is incorrect?

File: packages/api-plugin-promotions/src/resolvers/Mutation/duplicatePromotion.js [hunk 2]
```diff
@@ -8,6 +8,6 @@
 export default async function duplicatePromotion(_, { input }, context) {
   const { promotionId, shopId } = input;
   await context.validatePermissions("reaction:legacy:promotions", "create", { shopId });
-  const duplicatePromotionResults = await context.mutations.duplicatePromotion(context, promotionId);
+  const duplicatePromotionResults = await context.mutations.duplicatePromotion(context, { shopId, promotionId });
```
Comment (vanpho93) [inline packages/api-plugin-promotions/src/resolvers/Mutation/duplicatePromotion.js:11]: Should update on `duplicatePromotion` mutation file too.

File: packages/api-plugin-promotions/src/watchers/setPromotionState.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+async function markCompleted(context) {
+  const { collections: { Promotions } } = context;
+  const shopTimes = await getCurrentShopTime(context);
+  let totalUpdated = 0;
+  for (const shop of Object.keys(shopTimes)) {
+    const shopTime = shopTimes[shop];
+    // eslint-disable-next-line no-await-in-loop
+    const { modifiedCount } = await Promotions.updateMany({
+      shopId: shop,
+      state: "active",
```
Comment (vanpho93) [inline packages/api-plugin-promotions/src/watchers/setPromotionState.js:57]: is this state correct?

File: packages/api-plugin-promotions/src/watchers/setPromotionState.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+async function markActive(context) {
+  const { collections: { Promotions } } = context;
+  const shopTimes = await getCurrentShopTime(context);
+  let totalUpdated = 0;
+  for (const shop of Object.keys(shopTimes)) {
+    const shopTime = shopTimes[shop];
+    // eslint-disable-next-line no-await-in-loop
+    const { modifiedCount } = await Promotions.updateMany({
+      shopId: shop,
+      state: "created",
```
Comment (vanpho93) [inline packages/api-plugin-promotions/src/watchers/setPromotionState.js:30]: is this state correct?

File: packages/api-plugin-promotions/src/watchers/setPromotionState.js (hunk 3 above)
Comment (brent-hoover) [reply packages/api-plugin-promotions/src/watchers/setPromotionState.js:57]: Yes, why do you think it's not?

File: packages/api-plugin-promotions/src/watchers/setPromotionState.js (hunk 4 above)
Comment (brent-hoover) [reply packages/api-plugin-promotions/src/watchers/setPromotionState.js:30]: I think you might be looking at the filter part of the statement and thinking it's the $set part?

**Review by vanpho93 — APPROVED** [review]

**Pull request merged**